xla/pjrt/c/pjrt_c_api_raw_buffer_extension.h

```diff
@@ -0,0 +1,132 @@
+/* Copyright 2025 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef XLA_PJRT_C_PJRT_C_API_RAW_BUFFER_EXTENSION_H_
+#define XLA_PJRT_C_PJRT_C_API_RAW_BUFFER_EXTENSION_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "xla/pjrt/c/pjrt_c_api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct PJRT_RawBuffer PJRT_RawBuffer;
+
+struct PJRT_RawBuffer_CreateRawAliasOfBuffer_Args {
+  size_t struct_size;
+  PJRT_Extension_Base* extension_start;
+  PJRT_Buffer* buffer;
+  PJRT_RawBuffer* raw_buffer;  // out
+};
+PJRT_DEFINE_STRUCT_TRAITS(PJRT_RawBuffer_CreateRawAliasOfBuffer_Args,
+                          raw_buffer);
+
+typedef PJRT_Error* PJRT_RawBuffer_CreateRawAliasOfBuffer(
+    PJRT_RawBuffer_CreateRawAliasOfBuffer_Args* args);
+
+struct PJRT_RawBuffer_Destroy_Args {
+  size_t struct_size;
+  PJRT_Extension_Base* extension_start;
+  PJRT_RawBuffer* buffer;
+};
+PJRT_DEFINE_STRUCT_TRAITS(PJRT_RawBuffer_Destroy_Args, buffer);
+
+// Frees the PJRT_RawBuffer.
+typedef PJRT_Error* PJRT_RawBuffer_Destroy(PJRT_RawBuffer_Destroy_Args* args);
+
+struct PJRT_RawBuffer_GetOnDeviceSizeInBytes_Args {
+  size_t struct_size;
+  PJRT_Extension_Base* extension_start;
+  PJRT_RawBuffer* buffer;
+  size_t on_device_size_in_bytes;  // out
+};
+PJRT_DEFINE_STRUCT_TRAITS(PJRT_RawBuffer_GetOnDeviceSizeInBytes_Args,
+                          on_device_size_in_bytes);
+
+// Gets the number of bytes of the buffer storage on the device
+typedef PJRT_Error* PJRT_RawBuffer_GetOnDeviceSizeInBytes(
+    PJRT_RawBuffer_GetOnDeviceSizeInBytes_Args* args);
+
+struct PJRT_RawBuffer_GetMemorySpace_Args {
+  size_t struct_size;
+  PJRT_Extension_Base* extension_start;
+  PJRT_RawBuffer* buffer;
+  PJRT_Memory* memory_space;  // out
+};
+PJRT_DEFINE_STRUCT_TRAITS(PJRT_RawBuffer_GetMemorySpace_Args, memory_space);
+
+// Gets the memory space that this buffer is attached to.
+typedef PJRT_Error* PJRT_RawBuffer_GetMemorySpace(
+    PJRT_RawBuffer_GetMemorySpace_Args* args);
+
+struct PJRT_RawBuffer_CopyRawDeviceToHost_Args {
+  size_t struct_size;
+  PJRT_Extension_Base* extension_start;
+  PJRT_RawBuffer* buffer;
+  void* dst;
```

Comment:
@pschuh does this pointer need to be pinned?

Reply:
The impl has a fast-path for pinned buffers.